protokube/pkg/protokube/vsphere_volume.go

```diff
@@ -0,0 +1,180 @@
+	}
+
+	for _, etcd := range etcdClusters {
+		mountPoint := vsphere.GetMountPoint(etcd.VolumeId)
+		localDevice, err := getDevice(mountPoint)
+		if err != nil {
+			return nil, err
+		}
+		vol := &Volume{
+			ID:          etcd.VolumeId,
+			LocalDevice: localDevice,
+			AttachedTo:  attachedTo,
+			Mountpoint:  mountPoint,
+			Status:      VolStatusValue,
+			Info: VolumeInfo{
+				Description: etcd.EtcdClusterName,
+			},
+		}
+
+		etcdSpec := &EtcdClusterSpec{
+			ClusterKey: etcd.EtcdClusterName,
+			NodeName:   etcd.EtcdNodeName,
+		}
+
+		var nodeNames []string
+		for _, member := range etcd.Members {
+			nodeNames = append(nodeNames, member.Name)
+		}
+		etcdSpec.NodeNames = nodeNames
+		vol.Info.EtcdClusters = []*EtcdClusterSpec{etcdSpec}
+		volumes = append(volumes, vol)
+	}
+	glog.V(4).Infof("Found volumes: %v", volumes)
+	return volumes, nil
+}
+
+func getDevice(mountPoint string) (string, error) {
+	if runtime.GOOS == "linux" {
+		cmd := "lsblk"
+		arg := "-l"
+		out, err := exec.Command(cmd, arg).Output()
+		if err != nil {
+			return "", err
+		}
+
+		if Containerized {
+			mountPoint = PathFor(mountPoint)
+		}
+		lines := strings.Split(string(out), "\n")
+		for _, line := range lines {
+			if strings.Contains(line, mountPoint) {
+				lsblkOutput := strings.Split(line, " ")
+				glog.V(4).Infof("Found device: %v ", lsblkOutput[0])
+				return "/dev/" + lsblkOutput[0], nil
+			}
+		}
+	} else {
+		return "", fmt.Errorf("Failed to find device. OS %v is not supported for vSphere.", runtime.GOOS)
+	}
+	return "", fmt.Errorf("No device has been mounted on mountPoint %v.", mountPoint)
+}
+
+func getVolMetadata() ([]vsphere.VolumeMetadata, error) {
+	rawData, err := ioutil.ReadFile(PathFor(VolumeMetaDataFile))
+
+	if err != nil {
+		return nil, err
+	}
+
+	return vsphere.UnmarshalVolumeMetadata(string(rawData))
+}
+
+func (v *VSphereVolumes) AttachVolume(volume *Volume) error {
+	// Currently this is a no-op for vSphere. The virtual disks should already be mounted on this VM.
+	glog.Infof("All volumes should already be attached. No operation done.")
+	return nil
+}
+
+func (v *VSphereVolumes) InternalIp() net.IP {
+	if machineIp == nil {
+		ip, err := getMachineIp()
+		if err != nil {
+			return ip
+		}
+		machineIp = ip
+	}
+	return machineIp
+}
+
+func getMachineIp() (net.IP, error) {
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+	for _, iface := range ifaces {
+		if iface.Flags&net.FlagUp == 0 {
+			continue // interface down
+		}
+		if iface.Flags&net.FlagLoopback != 0 {
+			continue // loopback interface
+		}
+		addrs, err := iface.Addrs()
+		if err != nil {
+			return nil, err
+		}
+		for _, addr := range addrs {
+			var ip net.IP
+			switch v := addr.(type) {
+			case *net.IPNet:
+				ip = v.IP
+			case *net.IPAddr:
+				ip = v.IP
+			}
+			if ip == nil || ip.IsLoopback() {
+				continue
+			}
+			ip = ip.To4()
+			if ip == nil {
+				continue // not an ipv4 address
+			}
+			return ip, nil
```

Comment:
There could be potentially multiple IP addresses. Should you be checking for IP address for NIC with MAC address container vmware vendor ID?